macro-rules/src/java_types.rs

```diff
@@ -0,0 +1,269 @@
+//! These macros take in a "java type description" and generate various Rust types to reflect it.
+//! These java type descriptions are each a token tree and they have the following format:
+//!
+//! Java scalar types are a single identifier
+//! * `int`
+//! * `short`
+//! * etc
+//!
+//! Java reference types are a `()`-token tree like:
+//! * `(class[$path] $javaty*)`, e.g., `(class[java::util::Vector] (class[java::lang::String))` for `Vector<String>`
+//! * `(array $javaty)`, e.g., `(array[(class[java::lang::String])])` for `String[]`
+//! * `(generic $name)` to reference a generic (possible captured) type, e.g., `(generic[T])`
+
+/// Generates an `impl Trait` expression that is used as the type of a method argument
+/// in a Rust function reflecting a Java method.
+#[macro_export]
+macro_rules! argument_impl_trait {
+    ($scalar:ident $(+ $lt:lifetime)?) => {
+        impl duchess::IntoScalar< duchess::plumbing::rust_ty!($scalar) > $(+ $lt)?
+    };
+
+    ($r:tt $(+ $lt:lifetime)?) => {
+        impl duchess::IntoJava< duchess::plumbing::rust_ty!($r) > $(+ $lt)?
+    };
+}
```

Comment:
for the macros that aren't single-pattern and pretty obvious, I think some usage examples would probably be helpful